homeassistant/components/zwave_js/__init__.py

```diff
@@ -36,69 +38,130 @@ def register_node_in_dev_reg(
+    dev_reg = await device_registry.async_get_registry(hass)
 
     # pylint: disable=fixme
 
-    async def async_on_initialized():
-        """Handle initial full state received."""
-        # TODO: signal entities to update availability state
-        LOGGER.info("Connection to Zwave JS Server initialized")
-        initialized.set()
-
-        # This should be moved to the Node Ready command or something
-        # And then listen to new nodes too.
-        dev_reg = await device_registry.async_get_registry(hass)
-        for node in client.driver.controller.nodes.values():
-            register_node_in_dev_reg(entry, dev_reg, client, node)
+    async def async_on_connect():
+        """Handle websocket is (re)connected."""
+        LOGGER.info("Connected to Zwave JS Server")
+        if initialized.is_set():
+            # update entity availability
+            async_dispatcher_send(hass, f"{DOMAIN}_connection_state")
 
     async def async_on_disconnect():
         """Handle websocket is disconnected."""
         LOGGER.info("Disconnected from Zwave JS Server")
-        # TODO: signal entities to update availability state
+        async_dispatcher_send(hass, f"{DOMAIN}_connection_state")
 
-    for component in PLATFORMS:
-        hass.async_create_task(
-            hass.config_entries.async_forward_entry_setup(entry, component)
+    async def async_on_initialized():
+        """Handle initial full state received."""
+        LOGGER.info("Connection to Zwave JS Server initialized.")
+        initialized.set()
+
+    @callback
+    def async_on_node_ready(node: ZwaveNode) -> None:
+        """Handle node ready event."""
+        LOGGER.debug("Processing node %s", node)
+
+        # register (or update) node in device registry
+        register_node_in_dev_reg(entry, dev_reg, client, node)
+
+        # run discovery on all node values and create/update entities
+        for disc_info in async_discover_values(node):
+            LOGGER.debug("Discovered entity: %s", disc_info)
+            async_dispatcher_send(hass, f"{DOMAIN}_add_{disc_info.platform}", disc_info)
+
+    @callback
+    def async_on_node_added(node: ZwaveNode) -> None:
+        """Handle node added event."""
+        LOGGER.debug("Node added: %s - waiting for it to become ready.", node.node_id)
+        # we only want to run discovery when the node has reached ready state,
+        # otherwise we'll have all kinds of missing info issues.
+        if node.ready:
+            async_on_node_ready(node)
+            return
+        # if node is not yet ready, register one-time callback for ready state
+        node.once(
+            "ready",
+            lambda event: async_on_node_ready(event["node"]),
         )
+        # we do submit the node to device registry so user has
+        # some visual feedback that something is (in the process of) being added
+        register_node_in_dev_reg(entry, dev_reg, client, node)
+
+    async def handle_ha_shutdown(event):
```

Comment:
Import `Event` from core.py.
```suggestion
    async def handle_ha_shutdown(event: Event) -> None:
```